File: shared/origami/src/origami/comm/latency.cpp
/*******************************************************************************
 *
 * MIT License
 *
 * Copyright 2026 AMD ROCm(TM) Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 *******************************************************************************/

#include "origami/comm/latency.hpp"

namespace origami::comm {

namespace {
/// ceil_div clamped to a floor of 1: every stage of the iteration walk must run
/// at least once, even when the dividend rounds down to zero.
std::size_t ceil_div_min1(std::size_t numerator, std::size_t denominator) {
  return std::max<std::size_t>(ceil_div(numerator, denominator), 1);
}
}  // namespace

iter_times_t compute_iter_times(const functional_unit_work_t& work,
                                const system_t& system,
                                double bw_per_wg,
                                int active_cus,
                                const heuristics_t& heur,
                                std::optional<primitive_t> primitive) {
  const hardware_t& hw           = system.gpu;
  const comm_hardware_t& comm_hw = system.fabric;
  const double CL                = static_cast<double>(hw.cacheline_bytes);
  iter_times_t t{};

  // ── VMEM ──
  const double vmem_total = static_cast<double>(work.vmem_read_instrs + work.vmem_write_instrs);
  if (hw.vmem_issue_rate > 0.0) t.vmem = vmem_total / hw.vmem_issue_rate;

  // ── TCP ──
  if (hw.tcp_bw > 0.0) {
    t.tcp = static_cast<double>(work.tcp_read_cl + work.tcp_write_cl) * CL / hw.tcp_bw;
  }

  // ── L2 (scaled by active CUs on this XCD) ──
  // L2/TCC is per-XCD, so contention is decided by how many active CUs land on
  // one XCD, not the whole device. Spreading active_cus evenly over num_xcd
  // gives the per-XCD occupancy that l2_bw_per_cu_scaled reclaims idle slices
  // from.
  const int active_per_xcd =
      std::max<int>(static_cast<int>(std::ceil(static_cast<double>(active_cus) / hw.num_xcd)), 1);
  const double l2_bw = hw.l2_bw_per_cu_scaled(active_per_xcd);
  if (l2_bw > 0.0) { t.l2 = static_cast<double>(work.l2_read_cl + work.l2_write_cl) * CL / l2_bw; }

  // ── MALL ──
  if (hw.mall_bw > 0.0) {
    t.mall = static_cast<double>(work.mall_read_cl + work.mall_write_cl) * CL / hw.mall_bw;
  }

  // ── HBM (polynomial-scaled by active CUs) ──
  const double hbm_r_per_cu = hw.hbm_read_bw_per_cu(active_cus);
  const double hbm_w_per_cu = hw.hbm_write_bw_per_cu(active_cus);
  if (work.hbm_read_cl > 0 && hbm_r_per_cu > 0.0) {
    t.hbm_read = static_cast<double>(work.hbm_read_cl) * CL / hbm_r_per_cu;
  }
  if (work.hbm_write_cl > 0 && hbm_w_per_cu > 0.0) {
    t.hbm_write = static_cast<double>(work.hbm_write_cl) * CL / hbm_w_per_cu;
  }

  // ── xGMI read: latency-bound by outstanding-request limit ──
  // A remote read is gated by Little's law, not raw link width: a wave can
  // only keep mshr_depth misses in flight, and each takes xgmi_latency_cycles
  // (the 660 ns RTT) to return. So the bandwidth a WG can *sustain* is
  //   (in-flight bytes) / (round-trip latency)
  //   = (mshr_depth × waves_per_wg × CL) / xgmi_latency_cycles.
  // The WG cannot exceed either this latency cap or its share of the physical
  // link (bw_per_wg), so the effective rate is the min of the two.
  const double mshr_limited_bw =
      (static_cast<double>(hw.mshr_depth_per_wave) * hw.waves_per_wg * CL) / hw.xgmi_latency_cycles;
  const double effective_remote_read_bw = std::min(bw_per_wg, mshr_limited_bw);
  if (work.xgmi_read_cl > 0 && effective_remote_read_bw > 0.0) {
    t.xgmi_read = static_cast<double>(work.xgmi_read_cl) * CL / effective_remote_read_bw;
  }

  // ── xGMI write: concentration-limited ──
  // Writes behave differently from reads: a link is poorly utilized by a
  // single WG and only approaches full payload rate as more WGs pile onto it
  // (their in-flight stores overlap to hide framing/turnaround). Empirically
  // the link utilization follows a saturating ramp util = 1 − exp(−wgs/k):
  // ~1 WG reaches a small fraction, several WGs most of the link (k, set per
  // primitive in heuristics, controls how fast it saturates). We estimate how
  // many WGs share this link (link_bw / bw_per_wg), apply the ramp to get the
  // effective aggregate link bandwidth, then split it back per WG.
  if (work.xgmi_write_cl > 0 && bw_per_wg > 0.0) {
    const double wgs_on_link = std::max(comm_hw.link_bw / bw_per_wg, 1.0);
    // No operation context (nullopt) falls back to the default k, matching the
    // old empty-string behaviour of the string-keyed lookup.
    const double k =
        primitive ? heur.k_xgmi_write(*primitive) : heur.xgmi_write_concentration_k_default;
    const double util                = 1.0 - std::exp(-wgs_on_link / k);
    const double eff_link_bw         = comm_hw.link_bw * util;
    const double eff_write_bw_per_wg = eff_link_bw / wgs_on_link;
    t.xgmi_write = static_cast<double>(work.xgmi_write_cl) * CL / eff_write_bw_per_wg;
  }

  // ── VALU ──
  if (hw.valu_rate > 0.0 && work.valu_ops > 0) {
    t.valu = static_cast<double>(work.valu_ops) / hw.valu_rate;
  }

  return t;
}

std::pair<std::size_t, std::size_t> iter_counts_from_tile(const wg_tile_geometry_t& geometry,
                                                          std::size_t cl_per_iter,
                                                          std::size_t cacheline_bytes) {
  const std::optional<tile_shape_t>& shape = geometry.shape;

  // Strided tile: each of the m rows is walked on its own (a row's partial final
  // line can't merge with the next row), so iters = m * ceil(cl_per_row / cl_per_iter)
  // and each iteration reduces one row's-worth of columns spread over its iters_per_row.
  if (shape.has_value() && !shape->contiguous) {
    const std::size_t iters_per_row =
        ceil_div_min1(shape->cl_per_row(cacheline_bytes), cl_per_iter);
    const std::size_t num_iters         = std::max<std::size_t>(shape->m * iters_per_row, 1);
    const std::size_t elements_per_iter = ceil_div_min1(shape->n, iters_per_row);
    return {num_iters, elements_per_iter};
  }

  // Contiguous (or unknown) tile: one flat byte run chopped into cl_per_iter-line
  // iterations, with the elements split evenly across those iterations.
  const std::size_t num_iters         = ceil_div_min1(geometry.cachelines, cl_per_iter);
  const std::size_t elements_per_iter = ceil_div_min1(geometry.elements, num_iters);
  return {num_iters, elements_per_iter};
}

wg_tile_latency_breakdown_t compute_wg_tile_latency(const std::vector<op_t>& work_graph,
                                                    const wg_tile_geometry_t& geometry,
                                                    double bw_per_wg,
                                                    int active_cus,
                                                    const latency_context_t& ctx) {
  const hardware_t& hw           = ctx.system.gpu;
  const comm_hardware_t& comm_hw = ctx.system.fabric;
  const std::size_t cl_per_iter =
      static_cast<std::size_t>(ctx.config.cl_per_iter(hw.cacheline_bytes));
  const int instrs_per_cl = ctx.config.instrs_per_cl(hw.cacheline_bytes);

  auto [num_iters, elements_per_iter] =
      iter_counts_from_tile(geometry, cl_per_iter, hw.cacheline_bytes);

  const auto resolved = resolve_work_graph(
      work_graph,
      iter_dims_t{
          static_cast<int>(cl_per_iter), instrs_per_cl, static_cast<int>(elements_per_iter)});

  const iter_times_t times = compute_iter_times(
      resolved.iter_work, ctx.system, bw_per_wg, active_cus, ctx.heur, ctx.primitive);

  // Steady-state cost of one pipelined iteration = the binding FU (roofline).
  const double T_wlt = times.max_cycles();

  // Pipeline fill/drain. In steady state reads and writes overlap, but the
  // very first iteration's reads have nothing to hide behind (prologue) and
  // the very last iteration's writes have nothing following them (epilogue).
  // Each is the slowest stage on its side of the pipe: prologue = the dominant
  // inbound stage (local HBM read, remote xGMI read, or MALL), epilogue = the
  // dominant outbound stage (HBM or xGMI write).
  auto max_positive = [](std::initializer_list<double> xs) -> double {
    double best = 0.0;
    for (double v : xs)
      if (v > best) best = v;
    return best;
  };
  const double T_prologue = max_positive({times.hbm_read, times.xgmi_read, times.mall});
  const double T_epilogue = max_positive({times.hbm_write, times.xgmi_write});

  // One-time handshake for the whole tile (every signal/wait atomic), serial
  // with the transfer because the consumer cannot start until it is observed.
  const double T_sync =
      static_cast<double>(resolved.sync_work.atomic_count) * comm_hw.atomic_latency_cycles;

  // Fill + (num_iters−1) overlapped steady-state iterations + drain + sync.
  // The −1 is because the first iteration is already accounted for by the
  // prologue (its reads) and overlaps into the steady region.
  const double T_total = T_prologue +
                         std::max<double>(static_cast<double>(num_iters) - 1.0, 0.0) * T_wlt +
                         T_epilogue + T_sync;

  wg_tile_latency_breakdown_t out;
  out.T_total_cycles    = T_total;
  out.T_wlt_cycles      = T_wlt;
  out.T_prologue_cycles = T_prologue;
  out.T_epilogue_cycles = T_epilogue;
  out.T_sync_cycles     = T_sync;
  out.num_iters         = num_iters;

  out.T_vmem_cycles       = times.vmem;
  out.T_tcp_cycles        = times.tcp;
  out.T_l2_cycles         = times.l2;
  out.T_mall_cycles       = times.mall;
  out.T_hbm_read_cycles   = times.hbm_read;
  out.T_hbm_write_cycles  = times.hbm_write;
  out.T_xgmi_read_cycles  = times.xgmi_read;
  out.T_xgmi_write_cycles = times.xgmi_write;
  out.T_valu_cycles       = times.valu;
  out.bottleneck          = std::string{times.bottleneck()};

  out.clock_ghz = hw.clock_ghz;
  return out;
}

}  // namespace origami::comm

File: model/latency.py
"""
wg_tile latency computation. All times are in **GPU cycles**.

The atom (T_wlt): cycles for one WG to transfer one iter_tile over one
link, given all WGs contending for that link. The full wg_tile transfer
time is a software-pipelined loop of num_iters such atoms, mirroring
Origami's per-WG-tile latency structure:

    T_total = T_prologue + (num_iters - 1) × T_wlt + T_epilogue + T_sync

All `T_*` quantities here are cycle counts; the conversion to seconds
happens once at the top of the model (`predict_row` in collective.py).

`wg_tile_cachelines` is the one WG's share of a single gpu_timestep_tile
(see model/types.py for the full data hierarchy).
"""

from math import ceil, exp
from typing import List, Optional

from .types import (
    CommProblem, CommConfig, FunctionalUnitWork, TileShape,
    WgTileLatencyBreakdown, CACHELINE_BYTES, dtype_bytes,
)
from .hardware import Hardware, CommHardware
from .heuristics import Heuristics, DEFAULT_HEURISTICS
from .primitives import Op, Signal, Wait


def resolve_work_graph(
    ops: List[Op],
    cl_per_iter: int,
    instrs_per_cl: int,
    elements_per_iter: int,
) -> tuple[FunctionalUnitWork, FunctionalUnitWork]:
    """Resolve a work graph into per-iteration and sync work."""
    iter_work = FunctionalUnitWork.zero()
    sync_work = FunctionalUnitWork.zero()
    for op in ops:
        w = op.resolve(cl_per_iter, instrs_per_cl, elements_per_iter)
        if isinstance(op, (Signal, Wait)):
            sync_work += w
        else:
            iter_work += w
    return iter_work, sync_work


def compute_iter_times(
    work: FunctionalUnitWork,
    hw: Hardware,
    comm_hw: CommHardware,
    bw_per_wg: float,
    active_cus: int,
    heuristics: Heuristics = DEFAULT_HEURISTICS,
    primitive: Optional[str] = None,
) -> dict[str, float]:
    """Compute per-iteration time at each level of the hierarchy, in **cycles**.

    All hardware fields (`hw.*`, `comm_hw.*`) are cycle-based: bandwidths
    are bytes/cycle, latencies are cycles. The returned dict therefore
    holds cycle counts. Conversion to seconds happens once at the top
    of the model (`predict_row` in collective.py).

    Uses active_cus to scale HBM/L2 bandwidth via Origami's polynomial model
    (more CUs active → less per-CU bandwidth due to sharing).

    `heuristics` supplies the xGMI write concentration constant; the
    `primitive` name selects a per-collective k value if available.
    """
    CL = CACHELINE_BYTES

    vmem_total = work.vmem_read_instrs + work.vmem_write_instrs
    T_vmem = vmem_total / hw.vmem_issue_rate if hw.vmem_issue_rate > 0 else 0.0

    T_tcp = (work.tcp_read_cl + work.tcp_write_cl) * CL / hw.tcp_bw if hw.tcp_bw > 0 else 0.0

    # L2: scale by active CUs on this XCD
    active_per_xcd = max(ceil(active_cus / hw.num_xcd), 1)
    l2_bw = hw.l2_bw_per_cu_scaled(active_per_xcd)
    T_l2 = (work.l2_read_cl + work.l2_write_cl) * CL / l2_bw if l2_bw > 0 else 0.0

    T_mall = (work.mall_read_cl + work.mall_write_cl) * CL / hw.mall_bw if hw.mall_bw > 0 else 0.0

    # HBM: scale by active CUs via polynomial model
    hbm_r_per_cu = hw.hbm_read_bw_per_cu(active_cus)
    hbm_w_per_cu = hw.hbm_write_bw_per_cu(active_cus)
    T_hbm_read = work.hbm_read_cl * CL / hbm_r_per_cu if work.hbm_read_cl > 0 and hbm_r_per_cu > 0 else 0.0
    T_hbm_write = work.hbm_write_cl * CL / hbm_w_per_cu if work.hbm_write_cl > 0 and hbm_w_per_cu > 0 else 0.0

    # xGMI: limited by BOTH link bandwidth sharing AND MSHR depth (latency-bandwidth product)
    # Per-WG remote read throughput is min of:
    #   1. link_bw / wgs_on_link (bandwidth sharing) — bytes/cycle
    #   2. MSHR_depth × waves × cacheline / xGMI_latency (outstanding request limit)
    #      — bytes/cycle because xgmi_latency_cycles is in cycles
    mshr_limited_bw = (hw.mshr_depth_per_wave * hw.waves_per_wg * CL) / hw.xgmi_latency_cycles
    effective_remote_read_bw = min(bw_per_wg, mshr_limited_bw)
    T_xgmi_read = work.xgmi_read_cl * CL / effective_remote_read_bw if work.xgmi_read_cl > 0 and effective_remote_read_bw > 0 else 0.0

    # xGMI write: per-WG throughput scales with WG concentration on the
    # link via 1 - exp(-wgs/k). Empirically (rccl AG sweep, MI300X):
    #   1 WG/link → ~21% of payload, 2 → 38%, 5 → 65%, 9+ → 83%
    # The MSHR-cap-for-writes heuristic this replaces missed both the
    # single-WG floor (it overestimated by ~5x) and the multi-WG ramp.
    # See heuristics.xgmi_write_concentration_k and
    # wiki/pages/ag-outlier-investigation.md.
    #
    # Back out wgs_on_link from bw_per_wg — the layout has already
    # distributed WGs across links upstream:
    #   bw_per_wg = comm_hw.link_bw / wgs_on_link
    if work.xgmi_write_cl > 0 and bw_per_wg > 0:
        wgs_on_link = max(comm_hw.link_bw / bw_per_wg, 1.0)
        k = heuristics.k_xgmi_write(primitive)
        util = 1.0 - exp(-wgs_on_link / k)
        eff_link_bw = comm_hw.link_bw * util
        eff_write_bw_per_wg = eff_link_bw / wgs_on_link
        T_xgmi_write = work.xgmi_write_cl * CL / eff_write_bw_per_wg
    else:
        T_xgmi_write = 0.0

    T_valu = work.valu_ops / hw.valu_rate if hw.valu_rate > 0 and work.valu_ops > 0 else 0.0

    return {
        "vmem": T_vmem, "tcp": T_tcp, "l2": T_l2, "mall": T_mall,
        "hbm_read": T_hbm_read, "hbm_write": T_hbm_write,
        "xgmi_read": T_xgmi_read, "xgmi_write": T_xgmi_write,
        "valu": T_valu,
    }


def _iter_counts_from_tile(
    wg_tile: Optional[TileShape],
    wg_tile_cachelines: int,
    wg_tile_elements: int,
    cl_per_iter: int,
) -> tuple[int, int]:
    """Resolve (num_iters, elements_per_iter) for one wg_tile.

    Two cases when a 2D `wg_tile` is provided:

      - `contiguous=True` (default for dense per-rank tiles): the kernel
        streams the tile as a flat byte sequence — iteration count is the
        same as the 1D scalar path:
            num_iters         = ceil(cachelines / cl_per_iter)
            elements_per_iter = ceil(elements / num_iters)

      - `contiguous=False` (column stripe of a larger tile, post axis=1
        divide): the kernel walks each row separately and the partial-tail
        CL on every row costs a real iteration:
            iters_per_row     = ceil(cl_per_row / cl_per_iter)
            num_iters         = m × iters_per_row
            elements_per_iter = ceil(n / iters_per_row)

    For the degenerate 1D case (m=1) the contiguous-path collapses onto the
    scalar 1D answer, so existing CCL validation numbers are unaffected.
    """
    if wg_tile is not None and not wg_tile.contiguous:
        iters_per_row = max(ceil(wg_tile.cl_per_row / cl_per_iter), 1)
        num_iters = max(wg_tile.m * iters_per_row, 1)
        elements_per_iter = max(ceil(wg_tile.n / iters_per_row), 1)
        return num_iters, elements_per_iter
    # Contiguous tile (or no wg_tile supplied) — flat-byte iteration walk.
    num_iters = max(ceil(wg_tile_cachelines / cl_per_iter), 1)
    elements_per_iter = max(ceil(wg_tile_elements / num_iters), 1)
    return num_iters, elements_per_iter


def compute_wg_tile_latency(
    work_graph: List[Op],
    wg_tile_cachelines: int,
    config: CommConfig,
    hw: Hardware,
    comm_hw: CommHardware,
    bw_per_wg: float,
    wg_tile_elements: int,
    wg_tile: Optional[TileShape] = None,
    active_cus: Optional[int] = None,
    heuristics: Heuristics = DEFAULT_HEURISTICS,
    primitive: Optional[str] = None,
) -> WgTileLatencyBreakdown:
    """
    Compute the full wg_tile transfer latency for one timestep.

    T_wlt = one iter_tile = max(functional unit times) — the atom
    T_total = T_prologue + (num_iters - 1) × T_wlt + T_epilogue + T_sync

    If `wg_tile` (a 2D TileShape) is supplied, iteration counts are derived
    row-aware (m × ceil(cl_per_row / cl_per_iter)); otherwise the scalar
    `wg_tile_cachelines` / `wg_tile_elements` are used as a flat 1D chunk.

    `active_cus` controls the HBM/L2 bandwidth polynomial. When omitted,
    falls back to `config.num_wgs` (the requested parallelism); callers
    that know the effective WG count (e.g., after applying
    `CommConfig.effective_num_wgs`) should pass it explicitly so that
    over-launched WGs don't artificially depress per-CU bandwidth.
    """
    cl_per_iter = config.cl_per_iter
    instrs_per_cl = config.instrs_per_cl
    num_iters, elements_per_iter = _iter_counts_from_tile(
        wg_tile, wg_tile_cachelines, wg_tile_elements, cl_per_iter
    )

    iter_work, sync_work = resolve_work_graph(
        work_graph, cl_per_iter, instrs_per_cl, elements_per_iter
    )

    if active_cus is None:
        active_cus = config.num_wgs
    times = compute_iter_times(iter_work, hw, comm_hw, bw_per_wg, active_cus, heuristics, primitive)

    T_wlt = max(times.values()) if times else 0.0  # one iteration = one WLT

    read_times = [times["hbm_read"], times["xgmi_read"], times["mall"]]
    T_prologue = max(t for t in read_times if t > 0) if any(t > 0 for t in read_times) else 0.0

    write_times = [times["hbm_write"], times["xgmi_write"]]
    T_epilogue = max(t for t in write_times if t > 0) if any(t > 0 for t in write_times) else 0.0

    T_sync = sync_work.atomic_count * comm_hw.atomic_latency_cycles

    T_total = T_prologue + max(num_iters - 1, 0) * T_wlt + T_epilogue + T_sync

    bottleneck = max(times, key=times.get) if times else ""

    return WgTileLatencyBreakdown(
        T_total_cycles=T_total,
        T_wlt_cycles=T_wlt,
        T_prologue_cycles=T_prologue,
        T_epilogue_cycles=T_epilogue,
        T_sync_cycles=T_sync,
        num_iters=num_iters,
        T_vmem_cycles=times.get("vmem", 0),
        T_tcp_cycles=times.get("tcp", 0),
        T_l2_cycles=times.get("l2", 0),
        T_mall_cycles=times.get("mall", 0),
        T_hbm_read_cycles=times.get("hbm_read", 0),
        T_hbm_write_cycles=times.get("hbm_write", 0),
        T_xgmi_read_cycles=times.get("xgmi_read", 0),
        T_xgmi_write_cycles=times.get("xgmi_write", 0),
        T_valu_cycles=times.get("valu", 0),
        bottleneck=bottleneck,
        clock_ghz=hw.clock_ghz,
    )
